### PR TITLE
3732 upgrade common pydantic to v2

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes part of #3732 .

All changes related to common to upgrade pydantic to V2

More testing needs to be done as soon as this is merged into the base branch

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [ ] Do all end-to-end tests pass?
* [ ] Any other testing performed?
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [ ] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/common/agent_events/ram_consumption_event.py [hunk 1]
```diff
@@ -15,4 +16,4 @@ class RAMConsumptionEvent(AbstractAgentEvent):
     """
 
     utilization: Percent
-    bytes: conint(ge=0, strict=True)  # type: ignore [valid-type]
+    bytes: Annotated[int, Field(ge=0, strict=True)]  # type: ignore [valid-type]
```
Comment (mssalvatore) [inline monkey/common/agent_events/ram_consumption_event.py:19]: Do we still need `# type: ignore`?

File: monkey/common/agent_plugins/agent_plugin.py [hunk 2]
```diff
@@ -28,11 +23,6 @@ class AgentPlugin(InfectionMonkeyBaseModel):
...
-        # b64encode() returns bytes, so we call decode() to transform bytes to str
-        # Pydantic is not able to inherit AgentPluginManifest json encoders
-        # so we update the current encoders
-        json_encoders: Mapping[Type, Callable[[Any], Any]] = {
-            bytes: lambda byte_field: b64encode(byte_field).decode(),
-            **AgentPluginManifest.Config.json_encoders,
-        }
+    @field_serializer("source_archive", when_used="json")
+    def dump_source_archive(self, v):
+        return b64encode(v).decode()
```
Comment (mssalvatore) [inline monkey/common/agent_plugins/agent_plugin.py:28]: Is it still true that `Pydantic is not able to inherit AgentPluginManifest json encoders`? Do we still need this serializer?

File: monkey/common/agent_plugins/agent_plugin_repository_index.py [hunk 3]
```diff
@@ -26,6 +22,8 @@ class AgentPluginRepositoryIndex(MutableInfectionMonkeyBaseModel):
         :param plugins: Plugins' metadata, segregated by type and sorted by version
     """
 
+    model_config = ConfigDict(arbitrary_types_allowed=True, use_enum_values=True)
```
Comment (mssalvatore) [inline monkey/common/agent_plugins/agent_plugin_repository_index.py:25]: Shouldn't this be inheriting from `MutableInfectionMonkeyModelConfig`?

File: monkey/common/agent_plugins/agent_plugin_repository_index.py [hunk 4]
```diff
@@ -67,6 +63,7 @@ def _infection_monkey_version_parser(
 
         raise TypeError(f'Expected "{DEVELOPMENT}" or a valid semantic version, got {type(value)}')
 
-    @validator("plugins")
+    @field_validator("plugins")
+    @classmethod
     def _convert_str_type_to_enum(cls, plugins):
         return {AgentPluginType(t): plugins for t, plugins in plugins.items()}
```
Comment (mssalvatore) [inline monkey/common/agent_plugins/agent_plugin_repository_index.py:69]: Do we still need this?

File: monkey/tests/unit_tests/common/agent_events/test_credentials_stolen_events.py [hunk 5]
```diff
@@ -13,14 +13,14 @@
...
     assert PLAINTEXT_LM_HASH in serialized_event
     assert PLAINTEXT_NT_HASH in serialized_event
     assert PLAINTEXT_PRIVATE_KEY_1 in serialized_event
 
 
 def test_credential_stolen_event_deserialization_json():
-    serialized_event = TEST_EVENT.json()
-    deserialized_event = CredentialsStolenEvent.parse_raw(serialized_event)
+    serialized_event = TEST_EVENT.model_dump_json()
+    deserialized_event = CredentialsStolenEvent.model_validate_json(serialized_event)
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/common/agent_events/test_credentials_stolen_events.py:25]: We should add `to_json()` and `from_json()` to monkeytypes.

File: monkey/tests/unit_tests/common/agent_events/test_fingerprinting_event.py [hunk 6]
```diff
@@ -64,13 +64,13 @@ def test_from_dict():
...
 
-    assert fingerprinting_event.dict(simplify=True) == FINGERPRINTING_SIMPLE_DICT
+    assert fingerprinting_event.model_dump(mode="json") == FINGERPRINTING_SIMPLE_DICT
 
 
 def test_deserialization_dict():
     original = FINGERPRINTING_EVENT
 
-    serialized_event = original.dict()
+    serialized_event = original.model_dump()
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/common/agent_events/test_fingerprinting_event.py:73]: Consider add `to_dict()` to monkeytypes.

File: monkey/common/agent_plugins/agent_plugin_repository_index.py (hunk 3 above)
Comment (ilija-lazoroski) [reply monkey/common/agent_plugins/agent_plugin_repository_index.py:25]: Nope, UTs are in [monkeytypes](https://github.com/guardicode/monkeytypes/blob/dev/tests/test_base_models.py) about this

File: monkey/common/agent_plugins/agent_plugin_repository_index.py (hunk 4 above)
Comment (ilija-lazoroski) [reply monkey/common/agent_plugins/agent_plugin_repository_index.py:69]: Yep, we do. It will not be converted to enum if it is not there.

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**